src/sql/src/plan/statement.rs

```diff
@@ -753,35 +752,107 @@ fn handle_create_sink(scx: &StatementContext, stmt: Statement) -> Result<Plan, f
-            key_parts,
+            ref name,
+            ref on_name,
+            ref key_parts,
+            if_not_exists,
+        } => (
+            name.clone(),
+            on_name.clone(),
+            key_parts.clone(),
             if_not_exists,
-        } => (name, on_name, key_parts, if_not_exists),
+        ),
         _ => unreachable!(),
     };
     let on_name = scx.resolve_item(on_name)?;
     let catalog_entry = scx.catalog.get_item(&on_name);
-    let keys = query::plan_index_exprs(scx, catalog_entry.desc()?, &key_parts)?;
-    if !object_type_matches(ObjectType::View, catalog_entry.item_type())
-        && !object_type_matches(ObjectType::Source, catalog_entry.item_type())
+
+    if CatalogItemType::View != catalog_entry.item_type()
+        && CatalogItemType::Source != catalog_entry.item_type()
     {
         bail!(
             "index cannot be created on {} because it is a {}",
             on_name,
             catalog_entry.item_type()
-        );
+        )
     }
-    Ok(Plan::CreateIndex {
-        name: FullName {
+
+    let on_desc = catalog_entry.desc()?;
+
+    let filled_key_parts = match &key_parts {
+        Some(kp) => kp.to_vec(),
+        None => {
+            // `key_parts` is None if we're creating a "default" index, i.e.
+            // creating the index as if the index had been created alongside the
+            // view source, e.g. `CREATE MATERIALIZED...`
+            catalog_entry
+                .primary_idx_keys()?
+                .iter()
+                .map(|i| match on_desc.get_unambiguous_name(*i) {
+                    Some(n) => Expr::Identifier(vec![Ident::new(n.to_string())]),
+                    _ => Expr::Value(Value::Number((i + 1).to_string())),
+                })
+                .collect()
+        }
+    };
+    let keys = query::plan_index_exprs(scx, on_desc, &filled_key_parts)?;
+
+    let index_name = if let Some(name) = name {
+        FullName {
             database: on_name.database.clone(),
             schema: on_name.schema.clone(),
             item: normalize::ident(name),
-        },
+        }
+    } else {
+        let mut idx_name_base = on_name.clone();
+        if key_parts.is_none() {
+            // We're trying to create the "default" index.
+            idx_name_base.item += "_primary_idx";
+        } else {
+            // Use PG schema for automatically naming indexes:
+            // `<table>_<_-separated indexed cols>_idx`
+            let index_name_col_suffix = filled_key_parts.iter().map(|i| i.to_string()).join("_");
+            idx_name_base.item += &format!("_{}_idx", index_name_col_suffix);
+        }
+
+        let mut index_name = idx_name_base.clone();
+        let mut i = 0;
+
+        let mut cat_schema_iter = scx.catalog.list_items(&on_name.database, &on_name.schema);
+
+        // Search for an unused version of the name unless `if_not_exists`.
+        while cat_schema_iter.any(|i| *i.name() == index_name) && !if_not_exists {
```

Comment:
Apologies for the drive by review. This change looked cool and I wanted to check it out. I was trying to understand the logic here and having a tough time with it so I tried playing around with the code and found what I think may be a bug.

I created a view to play around with by doing:

```
materialize=> CREATE VIEW test (ts) AS VALUES (NULL), (TIMESTAMP '2040-01-01 10:10:10.111');
```

then I did this a bunch of times
```
materialize=> create default index on test;
```

eventually taking me to this state

```
materialize=> show indexes in test;
     Source_or_view      |               Key_name                | Column_name | Expression | Null | Seq_in_index
-------------------------+---------------------------------------+-------------+------------+------+--------------
 materialize.public.test | materialize.public.test_primary_idx   | ts          |            | t    |            1
 materialize.public.test | materialize.public.test_primary_idx1  | ts          |            | t    |            1
 materialize.public.test | materialize.public.test_primary_idx10 | ts          |            | t    |            1
 materialize.public.test | materialize.public.test_primary_idx2  | ts          |            | t    |            1
 materialize.public.test | materialize.public.test_primary_idx3  | ts          |            | t    |            1
 materialize.public.test | materialize.public.test_primary_idx4  | ts          |            | t    |            1
 materialize.public.test | materialize.public.test_primary_idx5  | ts          |            | t    |            1
 materialize.public.test | materialize.public.test_primary_idx6  | ts          |            | t    |            1
 materialize.public.test | materialize.public.test_primary_idx7  | ts          |            | t    |            1
 materialize.public.test | materialize.public.test_primary_idx8  | ts          |            | t    |            1
 materialize.public.test | materialize.public.test_primary_idx9  | ts          |            | t    |            1
(11 rows)
```

and now if I do it again I get

```
materialize=> create default index on test;
ERROR:  catalog item 'test_primary_idx10' already exists
materialize=>
```

Apologies in advance if this is not a real issue or I'm doing something not allowed here

Reply:
oof i see that as always, nikhil beat me to this realization :D sorry for the noise here sean

Reply:
np! thanks for checking it out. I did a bad copy and paste and deleted a very crucial iterator reassignment!